libraries/Microsoft.Bot.Builder.Ai.Translation/PostProcessor/PatternsPostProcessor.cs

```diff
@@ -0,0 +1,127 @@
+    /// and to handle words that needs to be kept same as source language from provided template each line having a regex
+    /// having first group matching the words that needs to be kept.
+    /// </summary>
+    public class PatternsPostProcessor : IPostProcessor
+    {
+        private readonly Dictionary<string, List<string>> _patterns;
+        private readonly Dictionary<string, HashSet<string>> _processedPatterns;
+
+        /// <summary>
+        /// Constructor that indexes input template for source language.
+        /// </summary>
+        /// <param name="patterns">No translate patterns for different languages</param> 
+        public PatternsPostProcessor(Dictionary<string, List<string>> patterns)
+        {
+            if(patterns == null)
+            {
+                throw new ArgumentNullException(nameof(patterns));
+            }
+
+            if(patterns.Count == 0)
+            {
+                throw new ArgumentException("Patterns can't be empty");
+            }
+
+            _processedPatterns = new Dictionary<string, HashSet<string>>();
+            foreach (KeyValuePair<string, List<string>> item in patterns)
+            {
+                _processedPatterns.Add(item.Key, new HashSet<string>());
+                foreach (string pattern in item.Value)
+                {
+                    string processedLine = pattern.Trim();
+                    if (!processedLine.Contains('('))
+                    {
+                        processedLine = '(' + processedLine + ')';
+                    }
+                    _processedPatterns[item.Key].Add(processedLine);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Process the logic for patterns post processor used to handle numbers and no translate list.
+        /// </summary>
+        /// <param name="translatedDocument">Translated document</param>
+        /// <param name="currentLanguage">Current source language</param>
+        /// <returns>A Task represents the asynchronus operation</returns>
+        public PostProcessedDocument Process(TranslatedDocument translatedDocument, string currentLanguage)
+        {
+            bool containsNum = Regex.IsMatch(translatedDocument.SourceMessage, @"\d");
+            string processedResult;
+            HashSet<string> temporaryPatterns = _processedPatterns[currentLanguage];
+            if (translatedDocument.LiteranlNoTranslatePhrases.Count > 0)
+            {
+                temporaryPatterns.UnionWith((translatedDocument.LiteranlNoTranslatePhrases));
+            }
+            if (temporaryPatterns.Count == 0 && !containsNum)
+                processedResult = translatedDocument.TargetMessage;
+            if (string.IsNullOrWhiteSpace(translatedDocument.RawAlignment))
+                processedResult = translatedDocument.TargetMessage;
+
+            var toBeReplaced = from result in temporaryPatterns
```

Comment:
I've tried really hard to avoid using LINQ statements like this. Is there a more clear way you could write this w/o LINQ? 

My experience (and the teams) is that these statements can be very hard to debug over time.